protostar/cairo_testing/cairo_test_execution_state.py

```diff
@@ -10,45 +11,164 @@
+            address_to_class_hash=self._contract_address_to_class_hash.copy(),
+            target_address_to_pranked_address=self._target_address_to_pranked_address.copy(),
+        )
+
+
+class BlockInfoControllerState:
+    def __init__(
+        self,
+        contract_address_to_block_timestamp: Optional[dict[Address, int]] = None,
+        contract_address_to_block_number: Optional[dict[Address, int]] = None,
+    ) -> None:
+        self._contract_address_to_block_timestamp = (
+            contract_address_to_block_timestamp or {}
+        )
+        self._contract_address_to_block_number = contract_address_to_block_number or {}
+
+    def get_block_timestamp(self, contract_address: Address):
+        return self._contract_address_to_block_timestamp.get(contract_address, None)
+
+    def get_block_number(self, contract_address: Address):
+        return self._contract_address_to_block_number.get(contract_address, None)
+
+    def set_block_timestamp(
+        self,
+        contract_address: Address,
+        block_timestamp: int,
+    ):
+        self._contract_address_to_block_timestamp[contract_address] = block_timestamp
+
+    def set_block_number(self, contract_address: Address, block_number: int):
+        self._contract_address_to_block_number[contract_address] = block_number
+
+    def remove_block_number(self, contract_address: Address):
+        del self._contract_address_to_block_number[contract_address]
+
+    def remove_block_timestamp(self, contract_address: Address):
+        del self._contract_address_to_block_timestamp[contract_address]
+
+    def clone(self):
+        return BlockInfoControllerState(
+            contract_address_to_block_number=self._contract_address_to_block_number.copy(),
+            contract_address_to_block_timestamp=self._contract_address_to_block_timestamp.copy(),
+        )
+
+
 @dataclass
 class CairoTestExecutionState:
-    starknet: Starknet
+    starknet_state: StarknetState
     stopwatch: Stopwatch
     output_recorder: OutputRecorder
     context: TestContext
     config: TestConfig
     project_compiler: ProjectCompiler
+    contracts_controller_state: ContractsControllerState
+    block_info_controller_state: BlockInfoControllerState
     expected_events_list: list[list[Event]] = field(default_factory=list)
 
-    @property
-    def cheatable_state(self) -> CheatableCachedState:
-        return cast(CheatableCachedState, self.starknet.state.state)
-
     def fork(self) -> Self:
         return dataclasses.replace(
             self,
             context=deepcopy(self.context),
             config=deepcopy(self.config),
             output_recorder=self.output_recorder.fork(),
             stopwatch=self.stopwatch.fork(),
-            starknet=self.starknet.copy(),
+            # pylint: disable=protected-access
+            starknet_state=self.starknet_state._copy(),
+            contracts_controller_state=self.contracts_controller_state.clone(),
```

Comment:
I believe it does not work properly. In the new architecture single test can execute multiple transactions, but you clone it only per test, not per transaction. We need copy apply mechanism implemented, thats why we used cached state before. Right now if you invoke transaction which reverts, protostar state may contain some lefovers from this transaciton. Does it make sense?

Reply:
It would make sense, if Protostar's hint local were available in a contract. However, they are not. 
That means contract execution doesn't mutate Protostar state. Changes to Protostar state are done after contract execution. Therefore, we don't need to implement the copy and apply mechanism.

```py
with self._starknet_state.copy_and_apply() as state_copy:
            call_info = await entry_point.execute_for_testing(
                state=state_copy,
                general_config=StarknetGeneralConfig(),
            )
            self._add_emitted_events(call_info.get_sorted_events()) # <= this won't execute in case of revert
```

Reply:
Yea, but we store there deployed contracts with syscalls as well